arangod/IResearch/IResearchLink.cpp

```diff
@@ -915,30 +1088,88 @@ Result IResearchLink::commitUnsafeImpl(bool wait, CommitResult* code) {
 
       commitLock.lock();
     }
-
-    auto const lastCommittedTick = _lastCommittedTick;
-
+    _commitStageOne = true;
+    auto const lastCommittedTick = _lastCommittedTickStageOne;
+
+#if ARANGODB_ENABLE_MAINTAINER_MODE && ARANGODB_ENABLE_FAILURE_TESTS
+    TRI_IF_FAILURE("ArangoSearch::ThreeTransactionsMisorder") {
+      std::unique_lock<std::mutex> sync(_t3Failureync);
+      while (_t3NumFlushRegistered < 2) {
+        sync.unlock();
+        std::this_thread::sleep_for(100ms);
+        sync.lock();
+      }
+      _t3CommitSignal = true;
+      LOG_TOPIC("4cb66", DEBUG, TOPIC) << "Commit started";
+    }
+#endif
     try {
       // _lastCommittedTick is being updated in '_before_commit'
       *code = _dataStore._writer->commit() ? CommitResult::DONE
                                            : CommitResult::NO_CHANGES;
     } catch (...) {
       // restore last committed tick in case of any error
-      _lastCommittedTick = lastCommittedTick;
+      _lastCommittedTickStageOne = lastCommittedTick;
       throw;
     }
-
     if (CommitResult::NO_CHANGES == *code) {
       LOG_TOPIC("7e319", TRACE, TOPIC)
           << "no changes registered for arangosearch link '" << id()
-          << "' got last operation tick '" << _lastCommittedTick << "'";
+          << "' got last operation tick '" << _lastCommittedTickStageOne << "'";
 
       // no changes, can release the latest tick before commit
       impl.tick(lastTickBeforeCommit);
-      _lastCommittedTick = lastTickBeforeCommit;
+      _lastCommittedTickStageOne = lastTickBeforeCommit;
+      _lastCommittedTickStageTwo = lastTickBeforeCommit;
```

Comment:
Is it safe and correct here to set both ticks to the same value, or do they need to be distinguished?

Reply:
Here it should be correct. As "first stage" have done nothing and tick that was before commit even started is for sure processed. As even if there was a transaction in-flight it would be registered in the flush context. But as here there were no transactions - we could release at least `lastTickBeforeCommit` and we know for sure that it is safe so skip anything before it during recovery.